File: 2023_07_31_functional_dependency_injection/typescript_impl/src/registerVisitWithEnv.ts
import { Option, some, none, pipe, error, ok, Result, matchValue, isUnion } from "powerfp";
import { Id, Patient, Payment, PatientRegistrationInfoDto, RegisterVisitRequestDto, RegisterVisitResponseDto, RegistrationError, Visit, Slot } from "./domain";
import { IEmailServiceP, IExternalPaymentServiceP, ILoggerP, IP, IPatientRepositoryP, IPaymentRepositoryP, IScheduleRepositoryP, InMemoryEmailService, createEnvWithSlots } from "./interfaces";
import * as assert from "assert";

type FirstArg<T extends (...args: any) => any> = T extends (arg1: infer A, ...args: infer P) => infer R ? A : any;


function getPatientById(env: IPatientRepositoryP, patientId: Id): Result<Patient, RegistrationError> {
    const patient = env.patientRepository.getPatientById(patientId);
    switch (patient.type) {
        case "some": return ok(patient.value);
        case "none": return error("PatientNotFound");
    }
}




// type SkipFirst<T extends (...args: any) => any> = T extends (arg1: any, ...args: infer P) => infer R ? (...args: P) => R : any;
// env: { getPatientById: SkipFirst<typeof getPatientById> } & IPatientRepositoryP & ILoggerP,


function addOrUpdatePatient(env: FirstArg<typeof getPatientById> & IPatientRepositoryP & ILoggerP, patient: PatientRegistrationInfoDto)
    : Result<[Option<Id>, string], RegistrationError> {

    switch (patient.type) {
        case "Anonymous": return ok([none, patient.email]);
        case "LoggedIn": return getPatientById(env, patient.patientId).bind(p => ok([some(p.id), p.email]));
        case "AnonymousWithCreatingNewAccount":
            const patientId = env.patientRepository.addPatient({ id: "", name: patient.email, email: patient.email });
            return ok([some(patientId), patient.email]);
        case "LoggedInWithUpdatingEmail":
            return getPatientById(env, patient.patientId)
                .bind(p => {
                    env.patientRepository.updatePatient({ ...p, email: patient.email });
                    env.logger.info(`Patient email has been changed from '${p.email}' to '${patient.email}' `);
                    return ok([some(p.id), patient.email])
                });
    }
}

function createPayment(env: IExternalPaymentServiceP & IPaymentRepositoryP, visitId: string, price: number, email: string): string {
    const externalPaymentId = env.externalPaymentService.createPayment(visitId, price, email);
    let payment = { id: "", userEmail: email, price, visitId, externalPaymentId };
    env.paymentRepository.addPayment(payment)
    return externalPaymentId;
}


function registerVisit(
    env: FirstArg<typeof addOrUpdatePatient> & FirstArg<typeof createPayment> & IScheduleRepositoryP & IEmailServiceP,
    registration: RegisterVisitRequestDto): Result<RegisterVisitResponseDto, RegistrationError> {

    const slotO = env.scheduleRepository.getSlotById(registration.slotId);
    switch (slotO.type) {
        case "none": return error("SlotNotFound");
        case "some":
            if (slotO.value.serviceIds.indexOf(registration.serviceId) === -1) {
                return error("SlotServiceNotFound");
            } else {
                const slot = slotO.value;
                return addOrUpdatePatient(env, registration.patient)
                    .bind(([patientIdO, email]) => {
                        const visit: Visit = { id: "", slotId: slot.id, from: slot.from, to: slot.to, serviceId: registration.serviceId, unitId: slot.unitId, patientId: patientIdO, patientEmail: email }
                        const visitId = env.scheduleRepository.addVisit(visit);
                        if (slot.price === 0) {
                            const emailContent = "New visit created ... ";
                            env.emailService.sendEmail(emailContent, email)
                            return ok({ visitId, externalPaymentId: none }) as Result<RegisterVisitResponseDto, RegistrationError>;
                        } else {
                            const externalPaymentId = createPayment(env, visitId, slot.price, email);
                            const emailContent = `New visit created, please pay in 15 minutes here $'{externalPaymentId}' `;
                            env.emailService.sendEmail(emailContent, email);
                            return ok({ visitId, externalPaymentId: some(externalPaymentId) });
                        }
                    });
            }

    }
}


var appEnv = createEnvWithSlots();

var registration: RegisterVisitRequestDto = { slotId: "1", patient: { type: 'LoggedIn', patientId: "1" }, serviceId: "1" };

type R = ReturnType<typeof registerVisit>;

assert(matchValue(registerVisit(createEnvWithSlots(), { ...registration, slotId: "3" }), error("SlotNotFound") as R));
assert(matchValue(registerVisit(createEnvWithSlots(), { ...registration, serviceId: "3" }), error("SlotServiceNotFound") as R));
assert(matchValue(registerVisit(createEnvWithSlots(), { ...registration, patient: { type: "LoggedIn", patientId: "2" } }), error("PatientNotFound") as R));

assert(matchValue(registerVisit(createEnvWithSlots(), registration), ok({ visitId: "1", externalPaymentId: none }) as R));
assert(matchValue(registerVisit(createEnvWithSlots(), { ...registration, slotId: "2" }), ok({ visitId: "1", externalPaymentId: some("payment_1") }) as R));


appEnv = createEnvWithSlots();
assert(matchValue(
    registerVisit(appEnv, { ...registration, patient: { type: "LoggedInWithUpdatingEmail", patientId: "1", email: "najder.marcin@gmail.com" } }),
    ok({ visitId: "1", externalPaymentId: none }) as R));
assert((appEnv.emailService as InMemoryEmailService).store[0].email === "najder.marcin@gmail.com");
let patient = appEnv.patientRepository.getPatientById("1");
assert(isUnion(patient, "some") && patient.value.email === "najder.marcin@gmail.com");



// testing only subfunctions

type R2 = ReturnType<typeof addOrUpdatePatient>;

assert(matchValue(
    addOrUpdatePatient(createEnvWithSlots(), { type: "Anonymous", email: "marcin.najder@gmail.com" }),
    ok([none, "marcin.najder@gmail.com"]) as R2));

assert(matchValue(
    addOrUpdatePatient(createEnvWithSlots(), { type: "AnonymousWithCreatingNewAccount", email: "marcin.najder@gmail.com" }),
    ok([some("2"), "marcin.najder@gmail.com"]) as R2));

assert(matchValue(
    addOrUpdatePatient(createEnvWithSlots(), { type: "LoggedIn", patientId: "1" }),
    ok([some("1"), "marcin.najder@gmail.com"]) as R2));

assert(matchValue(
    addOrUpdatePatient(createEnvWithSlots(), { type: "LoggedIn", patientId: "2" }),
    error("PatientNotFound") as R2));

appEnv = createEnvWithSlots();
assert(matchValue(
    addOrUpdatePatient(appEnv, { type: "LoggedInWithUpdatingEmail", patientId: "1", email: "najder.marcin@gmail.com" }),
    ok([some("1"), "najder.marcin@gmail.com"]) as R2));

File: 2023_07_31_functional_dependency_injection/fsharp_impl/FunctionalDependencyInjection/FunctionalDependencyInjection/RegisterVisitWithEnv.fs
module RegisterVisitWithEnv

open Domain
open Services

let getPatientById (env: #IPatientRepositoryP) patientId =
    match env.PatientRepository.GetPatientById patientId with
    | None -> Error PatientNotFound
    | Some patient -> Ok(patient)

let addOrUpdatePatient env (patient: PatientRegistrationInfoDto) =
    match patient with
    | Anonymous email -> Ok(None, email)
    | LoggedIn patientId ->
        getPatientById env patientId |> Result.bind (fun patient -> Ok((Some patientId), patient.Email))
    | AnonymousWithCreatingNewAccount email ->
        let patientId = env.PatientRepository.AddPatient { Id = ""; Name = email; Email = email }
        // let emailContent = $"New patient account has been created from '{email}'"
        // (env :> IEmailServiceP).EmailService.SendEmail(emailContent, email)
        Ok(Some patientId, email)
    | LoggedInWithUpdatingEmail (patientId, email) ->
        getPatientById env patientId
        |> Result.bind (fun patient ->
            env.PatientRepository.UpdatePatient { patient with Email = email }
            (env: ILoggerP).Logger.Info $"Patient email has been changed from '{patient.Email}' to '{email}' "
            Ok((Some patientId), email))

let createPayment (env: #IExternalPaymentServiceP) visitId price email =
    let externalPaymentId = env.ExternalPaymentService.CreatePayment(visitId, price, email)
    let payment =
        { Id = ""
          UserEmail = email
          Price = price
          VisitId = visitId
          ExternalPaymentId = externalPaymentId }
    (env :> IPaymentRepositoryP).PaymentRepository.AddPayment payment |> ignore
    externalPaymentId

// todo: result {}
let registerVisit (env: #IScheduleRepositoryP) (registration: RegisterVisitRequestDto) =
    match env.ScheduleRepository.GetSlotById registration.SlotId with
    | None -> Error SlotNotFound
    | Some slot ->
        if not (Array.contains registration.ServiceId slot.ServiceIds) then
            Error SlotServiceNotFound
        else
            addOrUpdatePatient env registration.Patient
            |> Result.bind (fun (patientIdO, email) ->
                let visit =
                    { Id = ""
                      SlotId = slot.Id
                      From = slot.From
                      To = slot.To
                      ServiceId = registration.ServiceId
                      UnitId = slot.UnitId
                      PatientId = patientIdO
                      PatientEmail = email }
                let visitId = env.ScheduleRepository.AddVisit visit
                if slot.Price = 0m then
                    let emailContent = $"New visit created ... "
                    (env :> IEmailServiceP).EmailService.SendEmail(emailContent, email)
                    Ok { VisitId = visitId; ExternalPaymentId = None }
                else
                    let externalPaymentId = createPayment env visitId slot.Price email
                    let emailContent = $"New visit created, please pay in 15 minutes here '{externalPaymentId}' "
                    (env :> IEmailServiceP).EmailService.SendEmail(emailContent, email)
                    Ok { VisitId = visitId; ExternalPaymentId = Some externalPaymentId })



let tests () =

    let (===) actual expected = if actual = expected then () else failwithf "assertion failed: %A <> %A" actual expected

    let mutable appEnv = createEnvWithSlots ()

    let mutable registration = { SlotId = "1"; Patient = LoggedIn "1"; ServiceId = "1" }

    // testing whole scenarios
    registerVisit (createEnvWithSlots ()) { registration with SlotId = "3" } === Error SlotNotFound

    registerVisit (createEnvWithSlots ()) { registration with ServiceId = "3" } === Error SlotServiceNotFound

    registerVisit (createEnvWithSlots ()) { registration with Patient = LoggedIn "2" } === Error PatientNotFound

    registerVisit (createEnvWithSlots ()) registration === Ok { VisitId = "1"; ExternalPaymentId = None }

    registerVisit (createEnvWithSlots ()) { registration with SlotId = "2" }
    === Ok { VisitId = "1"; ExternalPaymentId = Some "payment_1" }

    appEnv <- createEnvWithSlots ()

    registerVisit appEnv { registration with Patient = LoggedInWithUpdatingEmail("1", "najder.marcin@gmail.com") }
    === Ok { VisitId = "1"; ExternalPaymentId = None }

    (appEnv.getService<IEmailService> () :?> InMemoryEmailService).Store[0].Email === "najder.marcin@gmail.com"

    appEnv.getService<IPatientRepository>().GetPatientById("1").Value.Email === "najder.marcin@gmail.com"


    // testing only subfunctions
    addOrUpdatePatient (createEnvWithSlots ()) (Anonymous "marcin.najder@gmail.com")
    === Ok(None, "marcin.najder@gmail.com")

    addOrUpdatePatient (createEnvWithSlots ()) (AnonymousWithCreatingNewAccount "marcin.najder@gmail.com")
    === Ok(Some "2", "marcin.najder@gmail.com")

    addOrUpdatePatient (createEnvWithSlots ()) (LoggedIn "1") === Ok(Some "1", "marcin.najder@gmail.com")

    addOrUpdatePatient (createEnvWithSlots ()) (LoggedIn "2") === Error PatientNotFound

    appEnv <- createEnvWithSlots ()

    addOrUpdatePatient appEnv (LoggedInWithUpdatingEmail("1", "najder.marcin@gmail.com"))
    === Ok(Some "1", "najder.marcin@gmail.com")
